### PR TITLE
Use official Kubernetes python client

### PR DESCRIPTION
**Pull request by danielfrg**
This PR changes the `objects.py`to use the official [python kubernetes client](https://github.com/kubernetes-incubator/client-python). Hopefully this makes k8s specs creation a little easier by adding a pure python dependency.

This also solves https://github.com/jupyterhub/kubespawner/issues/6 by adding a simple `nodeSelector` to the Pod spec.

I also tried to solve https://github.com/jupyterhub/kubespawner/issues/1 but using `/user/{userid}` as a path for liveliness return a 404 because of Hub redirects or something (not 100% sure) any other ideas on how to check status of the Notebook server?

### CODE REVIEW

**Comment by yuvipanda** [comment]
Thank you for working on this! This looks great, am reviewing it now.

File: kubespawner/objects.py [hunk 1]
```diff
@@ -75,63 +103,98 @@ def make_pod_spec(
...
+    api_client = ApiClient()
+
+    pod = V1Pod()
+    pod.kind = "Pod"
+    pod.api_version = "v1"
+
+    pod.metadata = V1ObjectMeta()
+    pod.metadata.name = name
+    pod.metadata.labels = {}
+    pod.metadata.labels.update({"name": name})
```
Comment (yuvipanda) [inline kubespawner/objects.py]: why not just have this be one line?

File: kubespawner/objects.py [hunk 2]
```diff
@@ -11,13 +33,15 @@ def make_pod_spec(
     run_as_uid,
     fs_gid,
     env,
+    node_selector,
```
Comment (yuvipanda) [inline kubespawner/objects.py]: Can you split out the additional features into separate commits (or even a separate PR) so it's easier to review? I'll leave commnets about them here too tho

File: kubespawner/objects.py (hunk 1 above)
Comment (yuvipanda) [inline kubespawner/objects.py]:
```python
pod.metadata.labels = labels.copy()
pod.metadata.labels['name'] = name
```

seems clearer and possibly faster?

File: kubespawner/objects.py [hunk 3]
```diff
@@ -75,63 +103,98 @@ def make_pod_spec(
...
+        env_ = V1EnvVar()
+        env_.name = key
+        env_.value = value
+        notebook_container.env.append(env_)
+    notebook_container.args = cmd
+    notebook_container.image_pull_policy = image_pull_policy
+    notebook_container.resources = V1ResourceRequirements()
+    notebook_container.resources.requests = {"cpu": cpu_guarantee, "memory": mem_guarantee}
+    notebook_container.resources.limits = {"cpu": cpu_limit, "memory": mem_limit}
+    # notebook_container.liveness_probe = V1Probe()
```
Comment (yuvipanda) [inline kubespawner/objects.py]: Let's get the commented code out? Also would be nice if it were a separate commit :)

File: kubespawner/spawner.py [hunk 4]
```diff
@@ -109,6 +108,26 @@ def _namespace_default(self):
...
+
+        Default is None and means it will be launched in any available Node.
+        
+        For example to match the Nodes that have a label of `disktype: ssd` use:
+            `{"disktype": "ssd"}`
+        """
+    )
+
+    working_dir = Unicode(
+        "/home",
```
Comment (yuvipanda) [inline kubespawner/spawner.py]: I think the default should be None, which should make it use whatever is specified in the Docker image (which is the current default). Should probably be called singleuser_working_dir as well (I've been trying to do that for all new properties)

File: kubespawner/spawner.py [hunk 5]
```diff
@@ -109,6 +108,26 @@ def _namespace_default(self):
         """
     ).tag(config=True)
 
+
+    node_selector = Dict(
```
Comment (yuvipanda) [inline kubespawner/spawner.py]: Can this be a separate commit / PR? We need to be careful to support both simple property based selectors and more complex ones...

File: kubespawner/objects.py (hunk 1 above)
Comment (danielfrg) [inline kubespawner/objects.py]: Yes, thats better. Done

File: kubespawner/objects.py (hunk 2 above)
Comment (danielfrg) [inline kubespawner/objects.py]: I have removed the `node_selector` things I will make another PR for that.

File: kubespawner/objects.py (hunk 3 above)
Comment (danielfrg) [inline kubespawner/objects.py]: Yes, i wanted to have it there temporally in case you wanted it on this PR. It's gone.

File: kubespawner/spawner.py (hunk 4 above)
Comment (danielfrg) [inline kubespawner/spawner.py]: Ah yes, I agree that its a better default. Also renamed the property to `singleuser_working_dir`.

**Comment by danielfrg** [comment]
I have fixed the tests and removed the `node_selector` logic and I will make a separate PR for that Issue.

You can probably Squash and Merge this one if you want to keep history clean about changes.

File: kubespawner/spawner.py [hunk 6]
```diff
@@ -109,6 +108,13 @@ def _namespace_default(self):
         """
     ).tag(config=True)
 
+    singleuser_working_dir = Unicode(
+        help="""
```
Comment (yuvipanda) [inline kubespawner/spawner.py]: set the default explicitly to None and also allow_none?

**Comment by yuvipanda** [comment]
(I'll take a more detailed look + test on Monday!)

File: kubespawner/spawner.py (hunk 6 above)
Comment (danielfrg) [inline kubespawner/spawner.py]: Done

File: kubespawner/objects.py [hunk 7]
```diff
@@ -75,63 +98,86 @@ def make_pod_spec(
...
+    notebook_container.image = image_spec
+    notebook_container.working_dir = working_dir
+    notebook_container.ports = []
+    port_ = V1ContainerPort()
+    port_.name = "notebook-port"
+    port_.container_port = port
+    notebook_container.ports.append(port_)
+    notebook_container.env = []
+    for key, value in env.items():
+        env_ = V1EnvVar()
```
Comment (yuvipanda) [inline kubespawner/objects.py]: This could just be:

```python
notebook_container.env = [V1EnvVar(k, v) for k, v in env.items()]
```

no? https://github.com/kubernetes-incubator/client-python/blob/84bf6cc9dc206cb3314a5b4d2fa1a9c2fa2c5137/kubernetes/client/models/v1_env_var.py#L24 says it takes the properties as params to init

File: kubespawner/spawner.py [hunk 8]
```diff
@@ -433,6 +442,15 @@ def _hub_connect_port_default(self):
         """
     )
 
+    user_storage_match_labels = Dict(
+        help="""
```
Comment (yuvipanda) [inline kubespawner/spawner.py]: What's the default for this? I'm trying to understand what exactly this feature does - does it allow binding to an existing PVC that isn't created by this? Under what circumstances would I use this?

File: kubespawner/spawner.py [hunk 9]
```diff
@@ -469,19 +487,6 @@ def get_pod_manifest(self):
         """
         Make a pod manifest that will spawn current user's notebook pod.
         """
-        # Add a hack to ensure that no service accounts are mounted in spawned pods
-        # This makes sure that we don't accidentally give access to the whole
-        # kubernetes API to the users in the spawned pods.
-        # See https://github.com/kubernetes/kubernetes/issues/16779#issuecomment-157460294
-        hack_volumes = [{
```
Comment (yuvipanda) [inline kubespawner/spawner.py]: I think this belongs here, and *not* in the make_pod_spec. We could add an option to allow service account access in here when it's here, and this also keeps the make_pod_spec agnostic about what we do. Someone has actually forked this to make this disabling optional, and that's easier to do when this is here.

File: kubespawner/spawner.py [hunk 10]
```diff
@@ -496,38 +501,39 @@ def get_pod_manifest(self):
...
-            self.mem_guarantee,
+            name=self.pod_name,
+            image_spec=self.singleuser_image_spec,
+            image_pull_policy=self.singleuser_image_pull_policy,
+            image_pull_secret=self.singleuser_image_pull_secrets,
+            port=self.port,
+            cmd=real_cmd,
+            run_as_uid=singleuser_uid,
+            fs_gid=singleuser_fs_gid,
+            env=self.get_env(),
```
Comment (yuvipanda) [inline kubespawner/spawner.py]: Passing named args is much cleaner, thank you for moving it!

File: kubespawner/objects.py [hunk 11]
```diff
@@ -75,63 +98,86 @@ def make_pod_spec(
...
-        }
-    }
+        pod.spec.image_pull_secrets = []
+        image_secret = V1LocalObjectReference()
+        image_secret.name = image_pull_secret
+        pod.spec.image_pull_secrets.append(image_secret)
+
+    pod.spec.containers = []
+
+    notebook_container = V1Container()
```
Comment (yuvipanda) [inline kubespawner/objects.py]: We could also construct notebook_container this way (passing params to __init__) since that supports it too (https://github.com/kubernetes-incubator/client-python/blob/84bf6cc9dc206cb3314a5b4d2fa1a9c2fa2c5137/kubernetes/client/models/v1_container.py). I'm not super attached to it here, though - because there's a lot more params.

File: kubespawner/spawner.py (hunk 8 above)
Comment (danielfrg) [inline kubespawner/spawner.py]: So by default its nothing so the default behavior should be the same.

No, it still creates a PVC spec as it was doing before. This new property basically it just passes the `matchLabels` field from the PVC spec so its possible to filter between existing PV. Lets say you have multiple PV you can select which one you want.

From: https://kubernetes.io/docs/concepts/storage/persistent-volumes/ - PersistentVolumeClaims section.

Comment (yuvipanda) [inline kubespawner/spawner.py]: Ah, so is it for the case where the PV is not dynamically created by using a StorageClass?

File: kubespawner/objects.py (hunk 11 above)
Comment (danielfrg) [inline kubespawner/objects.py]: Yes I agree, personally I don't like to pass 3+ params to a function.

File: kubespawner/spawner.py (hunk 10 above)
Comment (danielfrg) [inline kubespawner/spawner.py]: 👍

File: kubespawner/objects.py (hunk 7 above)
Comment (danielfrg) [inline kubespawner/objects.py]: Yes, thats good. Done

**Comment by danielfrg** [comment]
This changes I have only tested on minikube but I still yet to see something in k8s that doesn't work from minukube to GCP.

File: kubespawner/spawner.py (hunk 9 above)
Comment (danielfrg) [inline kubespawner/spawner.py]: Yes I agree with keeping `make_pod_spec` independent of the hack i just wanted to keep the `kubernetes_client` code in one place. I put it back now.

File: kubespawner/spawner.py (hunk 8 above)
Comment (yuvipanda) [inline kubespawner/spawner.py]: ok! I think  we should set the default explicitly, and explain why you should use this and when in the help. The current help text isn't sufficient, IMO.

Also in the future, it'd be nice to have these as a separate PR, rather than as part of one :) Makes it easier to review.

I'll do some testing and try to get this merged tomorrow!

Comment (yuvipanda) [inline kubespawner/spawner.py]: The more I think about it, the more I think we should take it out now. I think way to include this would be:

1. Write a 'LabelSelector' trait type, that can be used to express either matchLabels or matchExpressions
2. Use that for this and for nodeSelectors

I think that's the cleanest and most future compatible way of doing this. what do you think?

Comment (yuvipanda) [inline kubespawner/spawner.py]: It should probably also support expanding {username} and friends in the label values, to provide more targeted matching. Let's definitely get it out of this PR?

Comment (danielfrg) [inline kubespawner/spawner.py]: Ok, its gone now. Yes, it was for that case where you already had existing PVs. This is the way I have always used PV + PVC, more manual than the automatic creation.

**Comment by yuvipanda** [comment]
Thank you for the patch!

**Comment by yuvipanda** [comment]
I'm going to test this now with https://github.com/jupyterhub/helm-chart

**Comment by yuvipanda** [comment]
I filed https://github.com/jupyterhub/kubespawner/issues/43 with further options on how to move forward getting rid of our custom code in favor of the official client.
